sagemaker_processing/spark_distributed_data_processing/sagemaker-spark-processing.ipynb

```diff
@@ -0,0 +1,562 @@
+   "outputs": [],
+   "source": [
+    "%%writefile ./code/hello_py_spark_app.py\n",
+    "import argparse\n",
+    "import time\n",
+    "\n",
+    "# Import local module to test spark-submit--py-files dependencies\n",
+    "import hello_py_spark_udfs as udfs\n",
+    "from pyspark.sql import SparkSession, SQLContext\n",
+    "from pyspark.sql.functions import udf\n",
+    "from pyspark.sql.types import IntegerType\n",
+    "import time\n",
+    "\n",
+    "if __name__ == \"__main__\":\n",
+    "    print(\"Hello World, this is PySpark!\")\n",
+    "\n",
+    "    parser = argparse.ArgumentParser(description=\"inputs and outputs\")\n",
+    "    parser.add_argument(\"--input\", type=str, help=\"path to input data\")\n",
+    "    parser.add_argument(\"--output\", required=False, type=str, help=\"path to output data\")\n",
+    "    args = parser.parse_args()\n",
+    "    spark = SparkSession.builder.appName(\"SparkTestApp\").getOrCreate()\n",
+    "    sqlContext = SQLContext(spark.sparkContext)\n",
+    "\n",
+    "    # Load test data set\n",
+    "    inputPath = args.input\n",
+    "    outputPath = args.output\n",
+    "    salesDF = spark.read.json(inputPath)\n",
+    "    salesDF.printSchema()\n",
+    "\n",
+    "    salesDF.createOrReplaceTempView(\"sales\")\n",
+    "\n",
+    "    # Define a UDF that doubles an integer column\n",
+    "    # The UDF function is imported from local module to test spark-submit--py-files dependencies\n",
+    "    double_udf_int = udf(udfs.double_x, IntegerType())\n",
+    "\n",
+    "    # Save transformed data set to disk\n",
+    "    salesDF.select(\"date\", \"sale\", double_udf_int(\"sale\").alias(\"sale_double\")).write.json(outputPath)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%writefile ./code/hello_py_spark_udfs.py\n",
+    "def double_x(x):\n",
+    "    return x + x"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create a processing job with python file dependencies\n",
+    "\n",
+    "Then, you'll create a processing job where the additional python file dependencies are specified via the `submit_py_files` argument in the `run()` function. If your Spark application requires additional jar file dependencies, these can be specified via the `submit_jars` argument of the `run()` function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define job input/output URIs\n",
+    "timestamp_prefix = strftime(\"%Y-%m-%d-%H-%M-%S\", gmtime())\n",
+    "prefix = 'sagemaker/spark-preprocess-demo/' + timestamp_prefix\n",
+    "input_prefix_sales = prefix + '/input/sales'\n",
+    "output_prefix_sales = prefix + '/output/sales'\n",
+    "input_s3_uri = 's3://' + bucket + '/' + input_prefix_sales\n",
+    "output_s3_uri = 's3://' + bucket + '/' + output_prefix_sales\n",
```

Comment:
Favor `.format` or `f-strings` here and elsewhere, rather than string concatenation with `+`.

Reply:
Updated to use `.format()` everywhere